pywhy_stats/categorical.py

```diff
@@ -0,0 +1,293 @@
+        observed_xy, _, _ = np.histogram2d(X, Y, bins=(np.unique(X).size, np.unique(Y).size))
+        chi, p_value, dof, expected = stats.chi2_contingency(observed_xy, lambda_=lambda_)
+
+    # Step 2: If there are conditionals variables, iterate over unique states and do
+    #         the contingency test.
+    else:
+        import pandas as pd
+
+        chi = 0
+        dof = 0
+
+        Z = np.asarray(Z)
+        if Z.ndim == 1:
+            Z = Z.reshape(-1, 1)
+
+        # XXX: needed when converting to only numpy API
+        # Check if all elements are integers
+        if not np.issubdtype(Z.dtype, np.integer):
+            le = LabelEncoder()
+            for idx in range(Z.shape[1]):
+                Z[:, idx] = le.fit_transform(Z[:, idx])
+            # warn("Converting Z array to categorical array using scikit-learn's LabelEncoder.")
+
+        # check number of samples relative to degrees of freedom
+        # assuming no zeros
+        s_size = Z.shape[1]
+        n_samples = Z.shape[0]
+
+        levels_x = len(np.unique(X))
+        levels_y = len(np.unique(Y))
+        dof_check = (
+            (levels_x - 1)
+            * (levels_y - 1)
+            * np.prod([len(np.unique(Z[:, i])) for i in range(s_size)])
+        )
+        n_samples_req = 10 * dof_check
+        if n_samples < n_samples_req:
+            raise RuntimeError(
+                f"Not enough samples. {n_samples} is too small. Need {n_samples_req}."
+            )
+
+        # XXX: currently we just leverage pandas to do the grouping. This is not
+        # ideal since we do not want the reliance on pandas package, but we should refactor
+        # this to only use numpy efficiently
+        X_columns = ["X"]
+        Y_columns = ["Y"]
+        Z_columns = [f"Z{i}" for i in range(Z.shape[1])]
+        columns = X_columns + Y_columns + Z_columns
+        data = pd.DataFrame(np.column_stack((X, Y, Z)), columns=columns)
+
+        for z_state, df in data.groupby(Z_columns[0] if len(Z_columns) == 1 else Z_columns):
+            try:
+                # Note: The fill value is set to 1e-7 to avoid the following error:
+                # where there are not enough samples in the data, which results in a nan pvalue
+                sub_table_z = (
+                    df.groupby(X_columns + Y_columns).size().unstack(Y_columns, fill_value=1e-7)
+                )
+                c, _, d, _ = stats.chi2_contingency(sub_table_z, lambda_=lambda_)
+                chi += c
+                dof += d
+            except ValueError:
+                # If one of the values is 0 in the 2x2 table.
+                if isinstance(z_state, str):
+                    logging.info(f"Skipping the test X \u27C2 Y | Z={z_state}. Not enough samples")
+                else:
+                    z_str = ", ".join([f"{var}={state}" for var, state in zip(Z_columns, z_state)])
+                    logging.info(f"Skipping the test X \u27C2 Y | {z_str}. Not enough samples")
```

Comment:
This means the test is still running and finishing even if potentially the majority of the `z_state` caused an exception? Would this really lead to the correct result then?

Reply:
Sounds good, can you add this?